File: java-app/src/main/java/com/yk/training/javaapp/service/BookService.java
package com.yk.training.javaapp.service;

import static java.lang.String.format;

import com.yk.training.javaapp.model.Book;
import com.yk.training.javaapp.utils.TimeUtils;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.springframework.stereotype.Service;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Service
public class BookService {

  private static final Logger LOGGER = LogManager.getLogger(BookService.class);

  private static final String[] AUTHORS = {"Josh", "Alex", "Bob", "Richard", "Kevin"};
  private static final int MAX_PAGE_SIZE = 100;
  private static final AtomicInteger COUNTER = new AtomicInteger(0);
  private final List<Book> books = initBooks();

  private static List<Book> initBooksWithTimer() {
    return TimeUtils.measureTime("initBooks", BookService::initBooks);
  }

  private static Book generateBook() {
    int counterValue = COUNTER.incrementAndGet();
    String title = "Title-" + (System.currentTimeMillis() % 100) + "-" + counterValue;
    String isbn = "ISBN-" + (System.currentTimeMillis() % 1000) + "-" + counterValue;
    int authorIndex = counterValue % AUTHORS.length;
    String author = AUTHORS[authorIndex] + "-" + counterValue;
    return new Book(author, title, isbn, counterValue);
  }

  private static List<Book> initBooks() {
    int nBooks = 100000;
    List<Book> books = new ArrayList<>(nBooks);
    for (int i = 0; i < nBooks; i++) {
      Book book = generateBook();
      books.add(book);
    }
    return books;
  }

  public List<Book> getBooks(int page, int pageSize) {
    if (page < 0) {
      throw new IllegalArgumentException(format("Page: %s must be greater than or equal to 0.", page));
    }

    if (pageSize > MAX_PAGE_SIZE) {
      throw new IllegalArgumentException(format("pageSize: %s must not be greater than %s", pageSize, MAX_PAGE_SIZE));
    }

    int start = (page - 1) * pageSize;
    if (start >= books.size()) {
      return new ArrayList<>();
    }

    int end = start + pageSize;
    if (end > books.size()) {
      end = books.size();
    }

    List<Book> res = books.subList(start, end);
    LOGGER.info("getBooks(start: {}, end: {}, size: {})", start, end, res.size());
    return res;
  }

  public Book getBook(int id) {
    if (id < 0 || id >= books.size()) {
      throw new IllegalArgumentException(format("ID: %s must be greater than or equal to 0.", id));
    }
    LOGGER.info("getBook({})", id);
    return books.get(id);
  }
}

File: rust-app/src/services/book_service.rs
use crate::models::book::Book;

use std::sync::atomic::{AtomicU32, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

static COUNTER: AtomicU32 = AtomicU32::new(0);
const AUTHORS: [&str; 5] = ["Josh", "Alex", "Bob", "Richard", "Kevin"];
const MAX_BOOKS: u32 = 1000;

pub fn get_book(id: u32) -> Option<Book> {
    if id == 0 || id > MAX_BOOKS {
        return None;
    }
    Some(generate_book(id))
}

pub fn get_books(page: u32, page_size: u32) -> Vec<Book> {
    let page = page.max(1);
    let start = (page - 1) * page_size + 1;
    let end = (start + page_size - 1).min(MAX_BOOKS);

    (start..=end).map(generate_book).collect()
}

fn generate_book(id: u32) -> Book {
    let counter_value = COUNTER.fetch_add(1, Ordering::SeqCst) + 1;

    // Equivalent of System.currentTimeMillis()
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("Time went backwards")
        .as_millis();

    let title = format!("Title-{}-{}", millis % 100, counter_value);
    let isbn = format!("ISBN-{}-{}", millis % 1000, counter_value);

    let author_index = (counter_value as usize) % AUTHORS.len();
    let author = format!("{}-{}", AUTHORS[author_index], counter_value);

    Book {
        id, // use requested id here
        title,
        author,
        isbn,
    }
}